### PR TITLE
`Communication`: Support user mentions and channel references

### PR DESCRIPTION
**Pull request by nityanandaz**
Render user mention substrings, e.g., `[user]Full Name(login)[/user]`, as @ followed by the user's name, e.g., `@Full Name`.

Render channel references substrings, e.g., `[channel]random(1337)[/channel]`, as # followed by the channel's name, e.g., `#random`

### CODE REVIEW

**Review by terlan98 — COMMENTED** [review]
Looks good! I like the idea of using the visitor pattern 👍 

I found an issue, but I'm not sure if it existed before this PR. When an exercise mention is enclosed in square brackets, it's not displayed correctly. This can be solved in a follow-up since it's unlikely that the users would put square brackets around a mention.

| iOS | Web |
|--------|--------|
| ios |web|


I also added a small suggestion.

**Comment by nityanandaz** [comment]
Good find, it matched `[[modeling]Overriding Methods(/courses/219/exercises/3395)[/modeling]]` to `[[modeling]Overriding Methods(/courses/219/exercises/3395)[/modeling]`, notice the two leading `[`.

Then `[modeling` does not equal `modeling`.

File: Sources/ArtemisMarkdown/ArtemisMarkdown.swift [hunk 1]
```diff
@@ -8,112 +8,24 @@ public struct ArtemisMarkdownView: View {
...
-        return replacedInsTag
+        var input = string
+        for visit in [
+            RegexReplacementVisitors.exercises.visit(input:),
+            RegexReplacementVisitors.ins.visit(input:),
+            RegexReplacementVisitors.lectures.visit(input:),
+            RegexReplacementVisitors.members.visit(input:)
+        ] {
+            visit(&input)
+        }
```
[the remark this answers is not in the document]
Comment (nityanandaz) [reply Sources/ArtemisMarkdown/ArtemisMarkdown.swift]: 66ca26fcdf9e82730203ef53722e58610accdf59

File: Sources/ArtemisMarkdown/RegexReplacementVisitor.swift [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+                return "file-upload"
+            case .text:
+                return "text"
+            case .modeling:
+                return "uml"
+            }
+        }
+    }
+
+    static let exercises = RegexReplacementVisitor(
```
[the remark this answers is not in the document]
Comment (nityanandaz) [reply Sources/ArtemisMarkdown/RegexReplacementVisitor.swift:59]: 66ca26fcdf9e82730203ef53722e58610accdf59

**Review by terlan98 — APPROVED** [review]
Thanks for addressing my comment 👍

**Pull request merged**